### ISSUE TITLE
Where can I find the enhanced wav data for the eval data?

### ISSUE THREAD

**Issue by Taruguchi** [espnet/espnet#4240]
With espnet/egs2/vctk_noisy as a reference, I learned speech enhancement using the data I prepared.
I prepared "train", "dev(valid)", "eval(test)" folder, and when the learning finished, I found the inference results about data for "dev" folder only.
(in espnet/egs2/vctk_noisy/enh1/exp/enh_train_raw/)

Where is the inference results about data for "eval"?
Or, should I check it myself using trained model?

**Comment by Emrys365** [comment]
Could you share the arguments that you passed to enh.sh? Like [egs2/vctk_noisy/enh1/run.sh#L16-L30](https://github.com/espnet/espnet/blob/master/egs2/vctk_noisy/enh1/run.sh#L16-L30)

Also, could you check the log in `espnet/egs2/vctk_noisy/enh1/exp/enh_train_raw/enhanced_*` to verify whether they finished successfully?

**Comment by Taruguchi** [comment]
Thanks your comment.
I uploaded run.txt.
(run.sh couldn't be uploaded, so I changed extension)
I checked enh1/exp/enh_train_raw/enhanced_dev folder, and I found "enhanced_dev" folder and wav data.
I want to know where the results for the test data.

[run.txt](https://github.com/espnet/espnet/files/8425185/run.txt)

**Comment by Emrys365** [comment]
The arguments look okay:
```bash
sample_rate=16k

train_set=train
valid_set=dev
test_sets=eval

./enh.sh \
    --lang en \
    --train_set "${train_set}" \
    --valid_set "${valid_set}" \
    --test_sets "${test_sets}" \
    --fs ${sample_rate} \
    --ngpu 1 \
    --spk_num 1 \
    --local_data_opts "" \
    --enh_config ./conf/train.yaml \
    --use_dereverb_ref false \
    --use_noise_ref false \
    --max_wav_duration 300 \
    --inference_model "valid.loss.best.pth" \
    "$@"
```

If stage 7 finished successfully, the enhanced audios should be present under `exp/enh_train_raw/enhanced_dev/logdir/output.*/wavs/*/`.

Could you check the last few lines in `exp/enh_train_raw/enhanced_dev/logdir/enh_inference.*.log` to verify whether they successfully finished with status 0?

**Comment by Taruguchi** [comment]
The learning finished successfully and "enhanced_dev" folder wav made.
There are the wav data that enhanced validation data, not test data. 

I think, "valid_set" is validation data, and "test_sets" is test data, is this correct?
And I think, validation data was used training, so I should confirm the performance of the model using test data which wasn't used training , right?

Please let me know if my idea is wrong.

**Comment by Emrys365** [comment]
The existence of the `enhanced_dev` folder does not mean inference on the dev set is successfully. It only succeeds when all 32 subprocesses (say `inference_nj =32` in [enh.sh](https://github.com/espnet/espnet/blob/master/egs2/TEMPLATE/enh1/enh.sh#L37)) exit with status 0. Are you sure about that?

On the other hand, do you have the log of running `run.sh`?
> It should look like follows:
> ```log
> 2022-04-06T13:24:57 (enh.sh:185:main) ./enh.sh --audio_format wav --train_set train --valid_set dev --test_sets test --fs 16k --lang zh --ngpu 1 --local_data_opts --outdir aishell4_simu --enh_config conf/tuning/train_enh_beamformer_no_wpe.yaml --stage 6 --stop_stage 8
> 2022-04-06T13:25:14 (enh.sh:510:main) Stage 6: Enhancemnt Frontend Training: train_set=dump/raw/train, valid_set=dump/raw/dev
> 2022-04-06T13:25:14 (enh.sh:565:main) Generate 'exp/enh_train_enh_beamformer_no_wpe_raw/run.sh'. You can resume the process from stage 6 using this script
> 2022-04-06T13:25:14 (enh.sh:568:main) enh training started... log: 'exp/enh_train_enh_beamformer_no_wpe_raw/train.log'
> 2022-04-06 13:25:57,224 (launch:95) INFO: espnet-0.10.7a1/tools/venv/envs/espnet/bin/python3 espnet-0.10.7a1/espnet2/bin/launch.py --cmd 'slurm.pl --name exp/enh_train_enh_beamformer_no_wpe_raw/train.log' --log exp/enh_train_enh_beamformer_no_wpe_raw/train.log --ngpu 1 --num_nodes 1 --init_file_prefix exp/enh_train_enh_beamformer_no_wpe_raw/.dist_init_ --multiprocessing_distributed true -- python3 -m espnet2.bin.enh_train --train_data_path_and_name_and_type dump/raw/train/wav.scp,speech_mix,sound --train_data_path_and_name_and_type dump/raw/train/spk1.scp,speech_ref1,sound --train_data_path_and_name_and_type dump/raw/train/spk2.scp,speech_ref2,sound --valid_data_path_and_name_and_type dump/raw/dev/wav.scp,speech_mix,sound --valid_data_path_and_name_and_type dump/raw/dev/spk1.scp,speech_ref1,sound --valid_data_path_and_name_and_type dump/raw/dev/spk2.scp,speech_ref2,sound --train_shape_file exp/enh_stats_16k/train/speech_mix_shape --train_shape_file exp/enh_stats_16k/train/speech_ref1_shape --train_shape_file exp/enh_stats_16k/train/speech_ref2_shape --valid_shape_file exp/enh_stats_16k/valid/speech_mix_shape --valid_shape_file exp/enh_stats_16k/valid/speech_ref1_shape --valid_shape_file exp/enh_stats_16k/valid/speech_ref2_shape --fold_length 80000 --fold_length 80000 --fold_length 80000 --resume true --output_dir exp/enh_train_enh_beamformer_no_wpe_raw --config conf/tuning/train_enh_beamformer_no_wpe.yaml
> 2022-04-06 13:25:57,671 (launch:349) INFO: log file: exp/enh_train_enh_beamformer_no_wpe_raw/train.log
> 2022-04-06T16:09:41 (enh.sh:602:main) Stage 7: Enhance Speech: training_dir=exp/enh_train_enh_beamformer_no_wpe_raw
> 2022-04-06T16:09:41 (enh.sh:612:main) Generate 'exp/enh_train_enh_beamformer_no_wpe_raw/run_enhance.sh'. You can resume the process from stage 7 using this script
> 2022-04-06T16:09:42 (enh.sh:641:main) Enhancement started... log: 'exp/enh_train_enh_beamformer_no_wpe_raw/enhanced_dev/logdir/enh_inference.*.log'
> 2022-04-06T17:01:58 (enh.sh:641:main) Enhancement started... log: 'exp/enh_train_enh_beamformer_no_wpe_raw/enhanced_test/logdir/enh_inference.*.log'
> espnet-0.10.7a1/egs2/aishell4/enh1/utils/slurm.pl: 7 / 32 failed, log is in exp/enh_train_enh_beamformer_no_wpe_raw/enhanced_test/logdir/enh_inference.*.log
> ```

If you see something like `7 / 32 failed`, then the inference on the dev set didn't finish successfully.
As a result, the inference on the test set will not be started, and you cannot find the enhanced audios for the test set.

-----

> I think, "valid_set" is validation data, and "test_sets" is test data, is this correct?

Yes.

> And I think, validation data was used training, so I should confirm the performance of the model using test data which wasn't used training , right?

Yes, in research we often care more about the performance on the test data.

**Comment by Taruguchi** [comment]
> do you have the log of running run.sh?

I have `train.log`, and there are sentences indicating that learning was completed in this log, such as below.

~~~[0b5fbc34ebd0] 2022-04-02 13:11:58,644 (trainer:334) INFO: 83epoch results: [train] iter_time=4.038e-04, forward_time=0.046, si_snr_loss=-13.758, loss=-13.758, backward_time=0.122, optim_step_time=0.008, optim0_lr0=1.563e-05, train_time=0.183, time=41 minutes and 42.96 seconds, total_count=1132369, gpu_max_cached_mem_GB=3.779, [valid] si_snr_loss=-12.726, loss=-12.726, time=1 minute and 17.26 seconds, total_count=135871, gpu_max_cached_mem_GB=3.779, [att_plot] time=2.16 seconds, total_count=0, gpu_max_cached_mem_GB=3.779
[0b5fbc34ebd0] 2022-04-02 13:11:59,465 (trainer:380) INFO: There are no improvements in this epoch
[0b5fbc34ebd0] 2022-04-02 13:11:59,467 (trainer:436) INFO: The model files were removed: exp/enh_train_raw/82epoch.pth
[0b5fbc34ebd0] 2022-04-02 13:11:59,467 (reporter:421) INFO: [Early stopping] valid.loss has not been improved 5 epochs continuously. The training was stopped at 83epoch
# Accounting: time=213979 threads=1
# Ended (code 0) at Sat Apr  2 13:11:59 UTC 2022, elapsed time 213979 seconds``~~~

**Comment by Emrys365** [comment]
This is only the log for stage 6 (training).
My suggestion is rerunning stage 7 and checking the log if you cannot find the previous log.

**Comment by Taruguchi** [comment]
I tried learning again, and such logs were outputted.

`2022-04-07T17:28:02 (enh.sh:601:main) Stage 7: Enhance Speech: training_dir=exp/enh_train_raw
2022-04-07T17:28:02 (enh.sh:611:main) Generate 'exp/enh_train_raw/run_enhance.sh'. You can resume the process from stage 7 using this script
2022-04-07T17:28:02 (enh.sh:640:main) Enhancement started... log: 'exp/enh_train_raw/enhanced_dev/logdir/enh_inference.*.log'
bash: line 1: 10635 Killed                  ( python3 -m espnet2.bin.enh_inference --ngpu 0 --fs 16k --data_path_and_name_and_type dump/raw/dev/wav.scp,speech_mix,sound --key_file exp/enh_train_raw/enhanced_dev/logdir/keys.16.scp --train_config exp/enh_train_raw/config.yaml --model_file exp/enh_train_raw/valid.loss.best.pth --output_dir exp/enh_train_raw/enhanced_dev/logdir/output.16 --normalize_output_wav true ) 2>> exp/enh_train_raw/enhanced_dev/logdir/enh_inference.16.log >> exp/enh_train_raw/enhanced_dev/logdir/enh_inference.16.log
bash: line 1: 10626 Killed                  ( python3 -m espnet2.bin.enh_inference --ngpu 0 --fs 16k --data_path_and_name_and_type dump/raw/dev/wav.scp,speech_mix,sound --key_file exp/enh_train_raw/enhanced_dev/logdir/keys.2.scp --train_config exp/enh_train_raw/config.yaml --model_file exp/enh_train_raw/valid.loss.best.pth --output_dir exp/enh_train_raw/enhanced_dev/logdir/output.2 --normalize_output_wav true ) 2>> exp/enh_train_raw/enhanced_dev/logdir/enh_inference.2.log >> exp/enh_train_raw/enhanced_dev/logdir/enh_inference.2.log
run.pl: 2 / 32 failed, log is in exp/enh_train_raw/enhanced_dev/logdir/enh_inference.*.log
`
They mean, my inference was failed?

**Comment by Taruguchi** [comment]
There is `enh_inference.*.log` in `enhanced_dev` , and this log was outputted.

~~~# python3 -m espnet2.bin.enh_inference --ngpu 0 --fs 16k --data_path_and_name_and_type dump/raw/dev/wav.scp,speech_mix,sound --key_file exp/enh_train_raw/enhanced_dev/logdir/keys.1.scp --train_config exp/enh_train_raw/config.yaml --model_file exp/enh_train_raw/valid.loss.best.pth --output_dir exp/enh_train_raw/enhanced_dev/logdir/output.1 --normalize_output_wav true 
# Started at Thu Apr  7 17:28:02 UTC 2022
#
/root/espnet/tools/anaconda/envs/espnet/bin/python3 /root/espnet/espnet2/bin/enh_inference.py --ngpu 0 --fs 16k --data_path_and_name_and_type dump/raw/dev/wav.scp,speech_mix,sound --key_file exp/enh_train_raw/enhanced_dev/logdir/keys.1.scp --train_config exp/enh_train_raw/config.yaml --model_file exp/enh_train_raw/valid.loss.best.pth --output_dir exp/enh_train_raw/enhanced_dev/logdir/output.1 --normalize_output_wav true
2022-04-07 17:28:04,380 (enh_inference:104) INFO: Perform direct speech enhancement on the input
/root/espnet/tools/anaconda/envs/espnet/lib/python3.8/site-packages/torch/_tensor.py:575: UserWarning: floor_divide is deprecated, and will be removed in a future version of pytorch. It currently rounds toward 0 (like the 'trunc' function NOT 'floor'). This results in incorrect rounding for negative values.
To keep the current behavior, use torch.div(a, b, rounding_mode='trunc'), or for actual floor division, use torch.div(a, b, rounding_mode='floor'). (Triggered internally at  /opt/conda/conda-bld/pytorch_1623448278899/work/aten/src/ATen/native/BinaryOps.cpp:467.)
  return torch.floor_divide(self, other)
# Accounting: time=267 threads=1
# Ended (code 0) at Thu Apr  7 17:32:29 UTC 2022, elapsed time 267 seconds~~~

**Comment by Emrys365** [comment]
Okay, this is only `enh_inference.1.log`, which finished without error. But you seem to misunderstand what I suggested.

By running
```bash
$ ./run.sh  --stage 7 --stage 8 > /path/to/log_file
```
or 
```bash
$ . ./path.sh
$ ${train_cmd} /path/to/log_file ./run.sh --stage 7 --stage 8
```
what would you see in the log file `/path/to/log_file`?

**Comment by Emrys365** [comment]
> I tried learning again, and such logs were outputted.
> 
> ```
> ...
> run.pl: 2 / 32 failed, log is in exp/enh_train_raw/enhanced_dev/logdir/enh_inference.*.log
> ```
> They mean, my inference was failed?

Yes, indeed. That's the reason why the inference on the test set was not started.
`run.pl: 2 / 32 failed` means 2 of 32 subtasks failed.
You could find them by running
```bash
tail -n1 exp/enh_train_raw/enhanced_dev/logdir/enh_inference.*.log
```
and checking the log files where the exit staus isn't 0.

**Comment by Taruguchi** [comment]
I found two log files that exit status isn't 0.
(`enh_inference.2.log` and `enh_inference.16.log`)
Their exit status is `code 137`.
I wil  send `enh_inference.2.log` for reference.

Compared to the other 30 log files, these logs show that nothing has been executed after statement below.
~~~
INFO: Perform direct speech enhancement on the input
~~~

But I can't understand why. Can you find the reason from this log?

[enh_inference.2.log](https://github.com/espnet/espnet/files/8449023/enh_inference.2.log)

**Comment by Emrys365** [comment]
The error code 137 seems to be related to the out-of-memory issue.

If you ran all subprocesses on different machines (like a cluster setup), could you try rerunning the inference for JOB=2 and JOB=16 on other machines with larger memory?

If you ran all subprocesses on a single machine at the same time, could you try rerunning only JOB=2 and JOB=16, respectively?

**Comment by Taruguchi** [comment]
When I try the training of vcth_noisy, I got the out-of-memory error in my environment.
So, I changed "batch_size" in train.yaml to 4, and tried my learning.

You told me that the error code 137 means the out-of-memory, so I changed "batch_size" to 2.

I modified enh.sh (JOB=16), and when this learning finished, enh_inference.16.log ended with code 0.
I will continue to modify enh.sh (JOB=2) and check the training results, thank you.

**Comment by Taruguchi** [comment]
I'm sorry, I have one more question.
Is there a way to inference unknown data (voice and noise mixed) using the trained model?

In the training of vctk_noisy, "train", "validation", "test" data are prepared, and 
each voice, noise and mixed wav are prepared respectivery.

What I want to know is whether inference of other wav data (mixed wav only) that doesn't belong to prepared data is feasible.
Or, I have to prepare voice, noise and mixed wav data when I try inference?

**Comment by Taruguchi** [comment]
@Emrys365 
Hi, I will tell you I solved my problems.
The performance of my PC was low, and it caused the problem about code 137 (out-of-memory), and I solved this problem by changed multithread process of enhancement to single thread.
And, I could find the way of inference unknown data. I modified run.sh to process only stage 7.
